lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include <string>
+#include <thread>
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+
+#include "lte/gateway/c/core/oai/test/mock_tasks/mock_tasks.h"
+#include "lte/gateway/c/core/oai/test/spgw_task/spgw_test_util.h"
+#include "lte/gateway/c/core/oai/include/spgw_state.h"
+#include "lte/gateway/c/core/oai/tasks/sgw/spgw_state_converter.h"
+#include "lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h"
+
+extern "C" {
+#include "lte/gateway/c/core/oai/include/mme_config.h"
+#include "lte/gateway/c/core/oai/tasks/sgw/pgw_handlers.h"
+#include "lte/gateway/c/core/oai/include/sgw_context_manager.h"
+#include "lte/gateway/c/core/oai/tasks/sgw/sgw_defs.h"
+#include "lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.h"
+#include "lte/gateway/c/core/oai/include/spgw_config.h"
+}
+
+extern bool hss_associated;
+
+namespace magma {
+namespace lte {
+
+task_zmq_ctx_t task_zmq_ctx_main_spgw;
+
+static int handle_message(zloop_t* loop, zsock_t* reader, void* arg) {
+  MessageDef* received_message_p = receive_msg(reader);
+
+  switch (ITTI_MSG_ID(received_message_p)) {
+    default: { } break; }
+
+  itti_free_msg_content(received_message_p);
+  free(received_message_p);
+  return 0;
+}
+
+class SPGWAppInjectedStateProcedureTest : public ::testing::Test {
+  virtual void SetUp() {
+    // setup mock MME app task
+    mme_app_handler = std::make_shared<MockMmeAppHandler>();
+    itti_init(
+        TASK_MAX, THREAD_MAX, MESSAGES_ID_MAX, tasks_info, messages_info, NULL,
+        NULL);
+
+    // initialize configs
+    mme_config_init(&mme_config);
+    spgw_config_init(&spgw_config);
+    create_partial_lists(&mme_config);
+    mme_config.use_stateless = false;
+    hss_associated           = true;
+
+    task_id_t task_id_list[2] = {TASK_MME_APP, TASK_SPGW_APP};
+    init_task_context(
+        TASK_MAIN, task_id_list, 2, handle_message, &task_zmq_ctx_main_spgw);
+
+    std::thread task_mme_app(start_mock_mme_app_task, mme_app_handler);
+    task_mme_app.detach();
+
+    std::cout << "Running setup" << std::endl;
+    // initialize the SPGW task
+    spgw_app_init(&spgw_config, mme_config.use_stateless);
+
+    std::string homedir = std::string(getpwuid(getuid())->pw_dir);
```

Comment:
I'd recommend adding a comment here to specify that the state injection is applied here